lib/CL/pocl_binary.c

```diff
@@ -0,0 +1,615 @@
+    return NULL;
+  if (b.version != POCLCC_VERSION)
+    return NULL;
+  if (strncmp(b.pocl_id, POCLCC_STRING_ID, POCLCC_STRING_ID_LENGTH))
+    return NULL;
+  if (pocl_binary_get_device_id(device) != b.device_id)
+    return NULL;
+  return p;
+}
+
+int pocl_binary_check_binary(cl_device_id device, const unsigned char *binary)
+{
+  return (check_binary(device, binary) != NULL);
+}
+
+/*****************************************************************************/
+
+void pocl_binary_set_program_buildhash(cl_program program,
+                                         unsigned device_i,
+                                         const unsigned char *binary)
+{
+  pocl_binary b;
+  read_header(&b, binary);
+  memcpy(program->build_hash[device_i],
+         b.program_build_hash, sizeof(SHA1_digest_t));
+}
+
+cl_uint pocl_binary_get_kernel_count(unsigned char *binary)
+{
+  pocl_binary b;
+  read_header(&b, binary);
+  return b.num_kernels;
+}
+
+cl_int pocl_binary_get_kernel_names(unsigned char *binary,
+                                    char **kernel_names,
+                                    size_t num_kernels)
+{
+  pocl_binary b;
+  unsigned char *buffer = read_header(&b, binary);
+  assert(num_kernels == b.num_kernels);
+
+  uint64_t struct_size;
+
+  unsigned char *orig_buffer;
+  unsigned i, len;
+  for (i=0; i < num_kernels; i++)
+  {
+    orig_buffer = buffer;
+    BUFFER_READ(struct_size, uint64_t);
+    // skip tar_size & arginfo_size
+    buffer += sizeof(uint64_t) + sizeof(uint32_t);
+    BUFFER_READ_STR2(kernel_names[i], len);
+    kernel_names[i][len] = 0;
+    buffer = orig_buffer + struct_size;
+  }
+  return CL_SUCCESS;
+}
+
+/***********************************************************/
+
+static unsigned char* tar_file(char* path, size_t basedir_offset, unsigned char* buffer)
+{
+  char* content;
+  uint64_t fsize;
+  char* p = path + basedir_offset;
+  BUFFER_STORE_STR(p);
+  pocl_read_file(path, &content, &fsize);
+  BUFFER_STORE_STR2(content, fsize);
+  free(content);
+  return buffer;
+}
+
+static unsigned char* recursively_tar_path(char* path, size_t basedir_offset, unsigned char* buffer)
+{
+  struct stat st;
+  stat(path, &st);
+
+  if (S_ISREG(st.st_mode))
+    buffer = tar_file(path, basedir_offset, buffer);
+
+  if (S_ISDIR(st.st_mode))
+    {
+      DIR *d;
+      struct dirent *entry;
+      char subpath[POCL_FILENAME_LENGTH];
+
+      strncpy(subpath, path, POCL_FILENAME_LENGTH);
+      char* p = subpath + strlen(subpath);
+      *p++ = '/';
+      d = opendir(path);
+      while ((entry = readdir(d)))
+        {
+          if (strcmp(entry->d_name, ".")==0) continue;
+          if (strcmp(entry->d_name, "..")==0) continue;
+          strcpy(p, entry->d_name);
+          buffer = recursively_tar_path(subpath, basedir_offset, buffer);
+        }
+      closedir(d);
+    }
+
+  return buffer;
+}
+
+static unsigned char* simple_tar(cl_kernel kernel, unsigned device_i, unsigned char* buffer)
```

Comment:
Simple tar? What is the difference to above? Please add purpose comments to these functions.